### PR TITLE
Run arcade unit tests in helix.

### PR DESCRIPTION
**Pull request by alexperovich**

### CODE REVIEW

File: tests/UnitTests.proj [hunk 1]
```diff
@@ -0,0 +1,47 @@
+<Project Sdk="Microsoft.DotNet.Helix.Sdk" DefaultTargets="Test">
+  <PropertyGroup>
+    <HelixSource>pr/dotnet/arcade/master</HelixSource>
```
Comment (chcosta) [inline tests/UnitTests.proj]: Maybe

```
<HelixSource>pr/dotnet/arcade/$(Build.SourceBranch)</HelixSource>
```

File: tests/UnitTests.proj [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    <HelixType>test/product</HelixType>
```
Comment (chcosta) [inline tests/UnitTests.proj]: does `HelixType` have to end with a `/`?  I thought I heard that somewhere

File: tests/UnitTests.proj [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+  <PropertyGroup Condition="'$(HelixTargetQueue)' == 'Windows.10.Amd64'">
+    <HelixPreCommands>$(HelixPreCommands);dir /s %HELIX_CORRELATION_PAYLOAD%</HelixPreCommands>
+    <HelixPreCommands>$(HelixPreCommands);dir /s</HelixPreCommands>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <XUnitProject Include="..\src\**\*.Tests.csproj"/>
+  </ItemGroup>
+
+  <!-- Useless stuff to make Arcade SDK happy -->
```
Comment (chcosta) [inline tests/UnitTests.proj]: > Useless stuff to make Arcade SDK happy

wow ;)

File: tests/UnitTests.proj [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  <PropertyGroup>
+    <Language>msbuild</Language>
+  </PropertyGroup>
+  <Target Name="Pack"/>
```
Comment (chcosta) [inline tests/UnitTests.proj]: What's the reason you need to override this target rather than just setting `IsPackable=false`?

Comment (alexperovich) [reply tests/UnitTests.proj]: This project doesn't have a Pack target.

File: azure-pipelines.yml [hunk 5]
```diff
@@ -80,13 +84,29 @@ jobs:
           $(_OfficialBuildIdArgs)
         displayName: Windows Build / Publish
 
+      - powershell: eng\common\build.ps1
+          -test
+          -projects $(Build.SourcesDirectory)/tests/UnitTests.proj
+        displayName: Run Helix Tests
+        env:
```
Comment (jonfortescue) [inline azure-pipelines.yml]: people tend to complain about making these env variables bc it makes it harder to repro locally. anyway we can parameterize these in a script?

File: tests/UnitTests.proj (hunk 1 above)
Comment (jonfortescue) [reply tests/UnitTests.proj]: agree

File: tests/UnitTests.proj (hunk 2 above)
Comment (jonfortescue) [reply tests/UnitTests.proj]: I feel like he's right

File: tests/UnitTests.proj [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    <!-- <HelixTargetQueue Include="Windows.10.Amd64"/> -->
+  </ItemGroup>
+
+  <PropertyGroup Condition="'$(HelixTargetQueue)' == 'Ubuntu.1604.Amd64'">
+    <HelixPreCommands>$(HelixPreCommands);find $HELIX_CORRELATION_PAYLOAD -type f</HelixPreCommands>
+    <HelixPreCommands>$(HelixPreCommands);find . -type f</HelixPreCommands>
+  </PropertyGroup>
+
+  <PropertyGroup Condition="'$(HelixTargetQueue)' == 'Windows.10.Amd64'">
+    <HelixPreCommands>$(HelixPreCommands);dir /s %HELIX_CORRELATION_PAYLOAD%</HelixPreCommands>
```
Comment (jonfortescue) [inline tests/UnitTests.proj]: I'm assuming these are here bc we're testing arcade with arcade so this will make debugging this easier, right?

File: azure-pipelines.yml [hunk 7]
```diff
@@ -81,13 +83,31 @@ jobs:
...
         ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
           name: dnceng-linux-internal-temp
+      variables:
+      - ${{ if and(ne(variables['System.TeamProject'], 'public'), notin(variables['Build.Reason'], 'PullRequest')) }}:
+        - group: DotNet-HelixApi-Access
+      # else
+      - ${{ if or(eq(variables['System.TeamProject'], 'public'), in(variables['Build.Reason'], 'PullRequest')) }}:
+        - group: Helix Anonymous
+        - name: HelixApiAccessToken
+          value: $(BotAccount-dotnet-github-anon-kaonashi-bot-helix-token)
```
Comment (riarenas) [inline azure-pipelines.yml]: Let's not use this variable group anymore now that it's not needed.

Comment (alexperovich) [reply azure-pipelines.yml]: I thought I got rid of this, :(

File: tests/UnitTests.proj [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+  </PropertyGroup>
+
+  <ItemGroup>
+    <HelixTargetQueue Include="Ubuntu.1604.Amd64"/>
+    <!--
+      TODO:
+      The OS onboarding queues don't have the required client changes for the xunit runner to work yet
+      Uncoment this once those changes are updated in the os onboarding queues
+    -->
+    <!-- <HelixTargetQueue Include="Windows.10.Amd64"/> -->
```
Comment (riarenas) [inline tests/UnitTests.proj]: The missing changes that this comment mentions should be in prod now. You should be able to enable windows.10.amd64 now

File: eng/common/build.sh [hunk 9]
```diff
@@ -133,6 +133,9 @@ while [[ $# > 0 ]]; do
     /m:*)
       properties="$properties $1"
       ;;
+    /bl:*)
```
Comment (chcosta) [inline eng/common/build.sh]: why is this needed?  Tomas has pushed back (in the past) on permitting the binlog name to be configurable unless there's some real scenario where this is required.

File: azure-pipelines.yml [hunk 10]
```diff
@@ -81,13 +83,27 @@ jobs:
           $(_ValidateSdkArgs)
         displayName: Windows Build / Publish
 
+      - script: set
```
Comment (chcosta) [inline azure-pipelines.yml]: Is this needed?

File: azure-pipelines.yml [hunk 11]
```diff
@@ -81,13 +83,27 @@ jobs:
...
+
+      - powershell: eng\common\build.ps1
+          -test
+          -projects $(Build.SourcesDirectory)/tests/UnitTests.proj
```
Comment (chcosta) [inline azure-pipelines.yml]: If Helix is supposed to be a first class citizen of Arcade, then shouldn't the repo know what projects are test projects such that you could just add "-test" and Arcade does the right thing?

**Comment by markwilkie** [comment]
What's the latest on this @alexperovich ?

**Comment by alexperovich** [comment]
I just finished the final changes to get this working. I have validation builds running to make sure. Once those run successfully this will be completed.

File: eng/common/build.sh (hunk 9 above)
Comment (chcosta) [reply eng/common/build.sh]: Yeah, this was one of the vectors that led me to have a larger conversation about better integrating Helix.

I'm not sure what the alternative is, but might want to give Tomas a heads up about this one.

@markwilkie

Comment (markwilkie) [reply eng/common/build.sh]: This should (in theory) be fine because it's a separate log from the build log.  e.g. outside of Arcade SDK's domain....
